crates/core/tedge_actors/src/servers/requests.rs

```diff
@@ -0,0 +1,229 @@
+use crate::Actor;
+use crate::ChannelError;
+use crate::DynSender;
+use crate::LoggingReceiver;
+use crate::Message;
+use crate::MessageReceiver;
+use crate::RuntimeError;
+use crate::RuntimeRequest;
+use crate::Sender;
+use crate::Server;
+use async_trait::async_trait;
+use futures::channel::oneshot;
+use futures::StreamExt;
+use log::error;
+use std::fmt::Debug;
+use std::ops::ControlFlow;
+
+/// Wrap a request with a [Sender] to send the response to
+///
+/// Requests are sent to server actors using such envelopes telling where to send the responses.
+pub struct RequestEnvelope<Request, Response> {
+    pub request: Request,
+    pub reply_to: Box<dyn Sender<Response>>,
+}
+
+impl<Request: Debug, Response> Debug for RequestEnvelope<Request, Response> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.request.fmt(f)
+    }
+}
+
+/// A message box used by a client to request a server and await the responses.
+pub struct ClientMessageBox<Request, Response> {
+    sender: DynSender<RequestEnvelope<Request, Response>>,
+}
+
+impl<Request: Message, Response: Message> ClientMessageBox<Request, Response> {
+    /// Send the request and await for a response
+    pub async fn await_response(&mut self, request: Request) -> Result<Response, ChannelError> {
+        let (sender, receiver) = oneshot::channel::<Response>();
+        let reply_to = Box::new(Some(sender));
+        self.sender
+            .send(RequestEnvelope { request, reply_to })
+            .await?;
+        let response = receiver.await;
+        response.map_err(|_| ChannelError::ReceiveError())
+    }
+}
+
+/// A [Sender] used by a client to send requests to a server,
+/// redirecting the responses to another recipient.
+pub struct RequestSender<Request: 'static, Response: 'static> {
+    sender: DynSender<RequestEnvelope<Request, Response>>,
+    reply_to: DynSender<Response>,
+}
+
+impl<Request, Response> Clone for RequestSender<Request, Response> {
+    fn clone(&self) -> Self {
+        RequestSender {
+            sender: self.sender.sender_clone(),
+            reply_to: self.reply_to.sender_clone(),
+        }
+    }
+}
+
+#[async_trait]
+impl<Request: Message, Response: Message> Sender<Request> for RequestSender<Request, Response> {
+    async fn send(&mut self, request: Request) -> Result<(), ChannelError> {
+        let reply_to = self.reply_to.sender();
+        self.sender
+            .send(RequestEnvelope { request, reply_to })
+            .await
+    }
+}
+
+/// An actor that wraps a request-response server
+///
+/// Requests are processed in turn, leading either to a response or an error.
+pub struct ServerActor<S: Server> {
```

Comment:
I'm assuming that the `ServerActor` and `ConcurrentServerActor` in the `servers::actors` module will be deprecated in favour of the ones defined here actor later on. Or is there any genuine use case for those?

Reply:
This is a draft I forgot to remove.